Intermittent error when updating Capacity Providers on ECS Cluster

There is an intermittent failure when updating the capacity providers on an ECS cluster when the cluster status is not `ACTIVE`.

This was first seen during acceptance testing of the feature.

```
--- FAIL: TestAccAWSEcsCluster_CapacityProvidersUpdate (26.05s)
    testing.go:635: Step 2 error: errors during apply:
        
        Error: error changing ECS cluster capacity provider settings (arn:aws:ecs:us-west-2:*******:cluster/tf-acc-test-1227245030962262513): ClientException: Cluster was not ACTIVE.
        	status code: 400, request id: b6cf25c3-7d49-4fad-876d-387b2c2a688b
```

The provider should both:
* wait for the cluster to be in `ACTIVE` state on creation and update
* retry the above exception when updating

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* `aws_ecs_cluster`

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor documentation? For example:
--->

* #11151

--- a/aws/resource_aws_ecs_cluster.go
+++ b/aws/resource_aws_ecs_cluster.go
@@ -267,7 +267,25 @@ func resourceAwsEcsClusterUpdate(d *schema.ResourceData, meta interface{}) error
 			input.DefaultCapacityProviderStrategy = expandEcsCapacityProviderStrategy(d.Get("default_capacity_provider_strategy").(*schema.Set))
 		}
 
-		_, err := conn.PutClusterCapacityProviders(&input)
+		err := resource.Retry(timeoutUpdate, func() *resource.RetryError {
+			_, err := conn.PutClusterCapacityProviders(&input)
+			if err != nil {
+				if isAWSErr(err, ecs.ErrCodeClientException, "Cluster was not ACTIVE") {
+					return resource.RetryableError(err)
+				}
+				if isAWSErr(err, ecs.ErrCodeResourceInUseException, "") {
+					return resource.RetryableError(err)
+				}
+				if isAWSErr(err, ecs.ErrCodeUpdateInProgressException, "") {
+					return resource.RetryableError(err)
+				}
+				return resource.NonRetryableError(err)
+			}
+			return nil
+		})
+		if isResourceTimeoutError(err) {
+			_, err = conn.PutClusterCapacityProviders(&input)
+		}
 		if err != nil {
 			return fmt.Errorf("error changing ECS cluster capacity provider settings (%s): %s", d.Id(), err)
 		}

--- a/aws/resource_aws_ecs_cluster.go
+++ b/aws/resource_aws_ecs_cluster.go
@@ -15,9 +15,9 @@ import (
 )
 
 const (
-	timeoutCreate = 10 * time.Minute
-	timeoutDelete = 10 * time.Minute
-	timeoutUpdate = 10 * time.Minute
+	ecsClusterTimeoutCreate = 10 * time.Minute
+	ecsClusterTimeoutDelete = 10 * time.Minute
+	ecsClusterTimeoutUpdate = 10 * time.Minute
 )
 
 func resourceAwsEcsCluster() *schema.Resource {
@@ -137,7 +137,7 @@ func resourceAwsEcsClusterCreate(d *schema.ResourceData, meta interface{}) error
 
 	d.SetId(aws.StringValue(out.Cluster.ClusterArn))
 
-	if err = waitForEcsClusterActive(conn, clusterName, timeoutCreate); err != nil {
+	if err = waitForEcsClusterActive(conn, clusterName, ecsClusterTimeoutCreate); err != nil {
 		return err
 	}
 
@@ -243,7 +243,7 @@ func resourceAwsEcsClusterUpdate(d *schema.ResourceData, meta interface{}) error
 			return fmt.Errorf("error changing ECS cluster settings (%s): %s", d.Id(), err)
 		}
 
-		if err = waitForEcsClusterActive(conn, clusterName, timeoutUpdate); err != nil {
+		if err = waitForEcsClusterActive(conn, clusterName, ecsClusterTimeoutUpdate); err != nil {
 			return err
 		}
 	}
@@ -267,7 +267,7 @@ func resourceAwsEcsClusterUpdate(d *schema.ResourceData, meta interface{}) error
 			input.DefaultCapacityProviderStrategy = expandEcsCapacityProviderStrategy(d.Get("default_capacity_provider_strategy").(*schema.Set))
 		}
 
-		err := resource.Retry(timeoutUpdate, func() *resource.RetryError {
+		err := resource.Retry(ecsClusterTimeoutUpdate, func() *resource.RetryError {
 			_, err := conn.PutClusterCapacityProviders(&input)
 			if err != nil {
 				if isAWSErr(err, ecs.ErrCodeClientException, "Cluster was not ACTIVE") {
@@ -290,7 +290,7 @@ func resourceAwsEcsClusterUpdate(d *schema.ResourceData, meta interface{}) error
 			return fmt.Errorf("error changing ECS cluster capacity provider settings (%s): %s", d.Id(), err)
 		}
 
-		if err = waitForEcsClusterActive(conn, clusterName, timeoutUpdate); err != nil {
+		if err = waitForEcsClusterActive(conn, clusterName, ecsClusterTimeoutUpdate); err != nil {
 			return err
 		}
 	}
@@ -305,7 +305,7 @@ func resourceAwsEcsClusterDelete(d *schema.ResourceData, meta interface{}) error
 	input := &ecs.DeleteClusterInput{
 		Cluster: aws.String(d.Id()),
 	}
-	err := resource.Retry(timeoutDelete, func() *resource.RetryError {
+	err := resource.Retry(ecsClusterTimeoutDelete, func() *resource.RetryError {
 		_, err := conn.DeleteCluster(input)
 
 		if err == nil {
@@ -338,7 +338,7 @@ func resourceAwsEcsClusterDelete(d *schema.ResourceData, meta interface{}) error
 	stateConf := resource.StateChangeConf{
 		Pending: []string{"ACTIVE", "DEPROVISIONING"},
 		Target:  []string{"INACTIVE"},
-		Timeout: timeoutDelete,
+		Timeout: ecsClusterTimeoutDelete,
 		Refresh: refreshEcsClusterStatus(conn, clusterName),
 	}
 	_, err = stateConf.WaitForState()

--- a/aws/resource_aws_ecs_cluster.go
+++ b/aws/resource_aws_ecs_cluster.go
@@ -14,6 +14,12 @@ import (
 	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
 )
 
+const (
+	timeoutCreate = 10 * time.Minute
+	timeoutDelete = 10 * time.Minute
+	timeoutUpdate = 10 * time.Minute
+)
+
 func resourceAwsEcsCluster() *schema.Resource {
 	return &schema.Resource{
 		Create: resourceAwsEcsClusterCreate,
@@ -131,6 +137,10 @@ func resourceAwsEcsClusterCreate(d *schema.ResourceData, meta interface{}) error
 
 	d.SetId(aws.StringValue(out.Cluster.ClusterArn))
 
+	if err = waitForEcsClusterActive(conn, clusterName, timeoutCreate); err != nil {
+		return err
+	}
+
 	return resourceAwsEcsClusterRead(d, meta)
 }
 
@@ -220,6 +230,8 @@ func resourceAwsEcsClusterRead(d *schema.ResourceData, meta interface{}) error {
 func resourceAwsEcsClusterUpdate(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).ecsconn
 
+	clusterName := d.Get("name").(string)
+
 	if d.HasChange("setting") {
 		input := ecs.UpdateClusterSettingsInput{
 			Cluster:  aws.String(d.Id()),
@@ -229,6 +241,10 @@ func resourceAwsEcsClusterUpdate(d *schema.ResourceData, meta interface{}) error
 		_, err := conn.UpdateClusterSettings(&input)
 		if err != nil {
 			return fmt.Errorf("error changing ECS cluster settings (%s): %s", d.Id(), err)
+		}
+
+		if err = waitForEcsClusterActive(conn, clusterName, timeoutUpdate); err != nil {
+			return err
 		}
 	}
 
@@ -255,6 +271,10 @@ func resourceAwsEcsClusterUpdate(d *schema.ResourceData, meta interface{}) error
 		if err != nil {
 			return fmt.Errorf("error changing ECS cluster capacity provider settings (%s): %s", d.Id(), err)
 		}
+
+		if err = waitForEcsClusterActive(conn, clusterName, timeoutUpdate); err != nil {
+			return err
+		}
 	}
 
 	return nil
@@ -267,7 +287,7 @@ func resourceAwsEcsClusterDelete(d *schema.ResourceData, meta interface{}) error
 	input := &ecs.DeleteClusterInput{
 		Cluster: aws.String(d.Id()),
 	}
-	err := resource.Retry(10*time.Minute, func() *resource.RetryError {
+	err := resource.Retry(timeoutDelete, func() *resource.RetryError {
 		_, err := conn.DeleteCluster(input)
 
 		if err == nil {
@@ -297,32 +317,13 @@ func resourceAwsEcsClusterDelete(d *schema.ResourceData, meta interface{}) error
 	}
 
 	clusterName := d.Get("name").(string)
-	dcInput := &ecs.DescribeClustersInput{
-		Clusters: []*string{aws.String(clusterName)},
+	stateConf := resource.StateChangeConf{
+		Pending: []string{"ACTIVE", "DEPROVISIONING"},
+		Target:  []string{"INACTIVE"},
+		Timeout: timeoutDelete,
+		Refresh: refreshEcsClusterStatus(conn, clusterName),
 	}
-	var out *ecs.DescribeClustersOutput
-	err = resource.Retry(5*time.Minute, func() *resource.RetryError {
-		log.Printf("[DEBUG] Checking if ECS Cluster %q is INACTIVE", d.Id())
-		out, err = conn.DescribeClusters(dcInput)
-
-		if err != nil {
-			return resource.NonRetryableError(err)
-		}
-		if !ecsClusterInactive(out, clusterName) {
-			return resource.RetryableError(fmt.Errorf("ECS Cluster %q is not inactive", clusterName))
-		}
-
-		return nil
-	})
-	if isResourceTimeoutError(err) {
-		out, err = conn.DescribeClusters(dcInput)
-		if err != nil {
-			return fmt.Errorf("Error waiting for ECS cluster to become inactive: %s", err)
-		}
-		if !ecsClusterInactive(out, clusterName) {
-			return fmt.Errorf("ECS Cluster %q is still not inactive", clusterName)
-		}
-	}
+	_, err = stateConf.WaitForState()
 	if err != nil {
 		return fmt.Errorf("Error waiting for ECS cluster to become inactive: %s", err)
 	}
@@ -331,15 +332,32 @@ func resourceAwsEcsClusterDelete(d *schema.ResourceData, meta interface{}) error
 	return nil
 }
 
-func ecsClusterInactive(out *ecs.DescribeClustersOutput, clusterName string) bool {
-	for _, c := range out.Clusters {
-		if aws.StringValue(c.ClusterName) == clusterName {
-			if *c.Status == "INACTIVE" {
-				return true
+func waitForEcsClusterActive(conn *ecs.ECS, clusterName string, timeout time.Duration) error {
+	stateConf := resource.StateChangeConf{
+		Pending: []string{"PROVISIONING"},
+		Target:  []string{"ACTIVE"},
+		Timeout: timeout,
+		Refresh: refreshEcsClusterStatus(conn, clusterName),
+	}
+	_, err := stateConf.WaitForState()
+	return err
+}
+
+func refreshEcsClusterStatus(conn *ecs.ECS, clusterName string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		output, err := conn.DescribeClusters(&ecs.DescribeClustersInput{
+			Clusters: []*string{aws.String(clusterName)},
+		})
+		if err != nil {
+			return nil, "", err
+		}
+		for _, c := range output.Clusters {
+			if aws.StringValue(c.ClusterName) == clusterName {
+				return c, aws.StringValue(c.Status), nil
 			}
 		}
+		return nil, "", fmt.Errorf("ECS cluster %q missing", clusterName)
 	}
-	return false
 }
 
 func expandEcsSettings(configured *schema.Set) []*ecs.ClusterSetting {

